internal/util/sysreadfile.go

```diff
@@ -35,11 +39,30 @@ func SysReadFile(file string) (string, error) {
 	//
 	// Since we either want to read data or bail immediately, do the simplest
 	// possible read using syscall directly.
-	b := make([]byte, 128)
+	const sysFileBufferSize = 128
+	b := make([]byte, sysFileBufferSize)
 	n, err := syscall.Read(int(f.Fd()), b)
 	if err != nil {
 		return "", err
 	}
 
 	return string(bytes.TrimSpace(b[:n])), nil
 }
+
+// ReadFileNoStat uses ioutil.ReadAll to read contents of entire file.
+// This is similar to ioutil.ReadFile but without the call to os.Stat, because
+// many files in /proc and /sys report incorrect file sizes (either 0 or 4096).
+// Reads a max file size of 512kB.  For files larger than this, a scanner
+// should be used.
+func ReadFileNoStat(filename string) ([]byte, error) {
+	const maxBufferSize = 1024 * 500
+
+	f, err := os.Open(filename)
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+
+	reader := io.LimitReader(f, maxBufferSize)
+	return ioutil.ReadAll(reader)
+}
```

Comment:
This .go file has the following build constraint on line 14:

```
// +build linux,!appengine
```

If you're adding a new function here, you should also add it to [internal/util/sysreadfile_compat.go](https://github.com/prometheus/procfs/blob/c0c2a8be4d30a2e2cb95ea371a6f32a506d3e45e/internal/util/sysreadfile_compat.go) file. Otherwise, this function will only be available on non-appengine linux.

Reply:
Unlike `SysReadFile`, `ReadFileNoStat` doesn't actually use syscalls, so it should probably be added to another .go file without the build constraint. That way you wouldn't have to have 2 identical implementations of it in `sysreadfile.go` and `sysreadfile_compat.go`.